A few updates to Mike's first pass as a result of my review...

OliverPearmain-Pusher opened the pull request:
- Tried to get rid of any unwrapping and force of optionals where possible
- Improved naming of some variables (mostly in the cell classes)
- Altered some method signatures to be more swifty (e.g. `itemAt(:)` -> `item(at:)` )
- Remove `DispatchQueue.main.async` from `viewDidLoad` (this seemed unnecessary)
- Moved `Delegate` protocol definitions to top of file (this is where they typically live in my experience)
- Removed any unnecessary `public` access modifiers, added `private` where I could
- Combined the cell `configure` and `setImage` methods into a single `configure` and simplified the construction of cells
- Added `preconditionFailure` in event of invalid plist values
- Removed `MessageMapper` in favour of exposing the values as properties via extension on the originating entity

OliverPearmain-Pusher committed:
A few updates to Mike's first pass as a result of my review...
- Tried to get rid of any unwrapping and force of optionals where possible
- Improved naming of some variables (mostly in the cell classes)
- Altered some method signatures to be more swifty (e.g. `itemAt(:)` -> `item(at:)` )
- Remove `DispatchQueue.main.async` from `viewDidLoad` (this seemed unnecessary)
- Moved `Delegate` protocol definitions to top of file (this is where they typically live in my experience)
- Removed any unnecessary `public` access modifiers, added `private` where I could
- Combined the cell `configure` and `setImage` methods into a single `configure` and simplified the construction of cells
- Added `preconditionFailure` in event of invalid plist values
- Removed `MessageMapper` in favour of exposing the values as properties via extension on the originating entity

OliverPearmain-Pusher committed:
Removed redundant unwrap.

mdpye commented on app/Chatkit Quickstart/ChatroomViewController.swift:
```diff
@@ -58,96 +54,101 @@ class ChatroomViewController: UIViewController {
                 return
             }
-            
-            DispatchQueue.main.async {
```
I think this is still required, the callback won't be made on the main thread, and even if we don't touch any UI components directly, none of the controller methods are really threadsafe...

grzesiekko replied:
This sounds sensible, especially knowing how highly concurrent and barely synched v1 is.

OliverPearmain-Pusher replied:
I have now added the `DispatchQueue.main.async` back :-)

mdpye commented:
This is great, thanks

mdpye commented:
Any comments on how easy the overall project is to follow? (assuming many comments are added)

grzesiekko commented on app/Chatkit Quickstart/MessagesDataModel.swift:
```diff
@@ -18,16 +21,45 @@ struct LocalMessage {
 }
 
+protocol MessagesDataModelDelegate {
+    func didChange(model: MessagesDataModel.MessagesModel, changeType: ChangeType)
```
```suggestion
    func messagesDataModel(_ messagesDataModel: MessagesDataModel, didAddMessagesModel messagesModel: MessagesDataModel.MessagesModel, atIndex index: Int)
    func messagesDataModel(_ messagesDataModel: MessagesDataModel, didUpdateMessagesModel messagesModel: MessagesDataModel.MessagesModel, atIndex index: Int)
```

OliverPearmain-Pusher replied:
Good suggestion.  I have committed this change as discussed.

grzesiekko commented on app/Chatkit Quickstart/SenderMessageTableViewCell.swift:
```diff
@@ -2,43 +2,53 @@ import UIKit
+        if let senderAvatarUrlString = senderAvatarUrl,
+            let senderAvatarUrl = URL(string: senderAvatarUrlString) {
+    
+            URLSession.shared.dataTask(with: senderAvatarUrl) { (data, _, _) in
```
One could argue that it is not the view responsibility to download an image. However, I do agree that it is convenient to do it here.

OliverPearmain-Pusher replied:
I agree but I guess we're trying to keep this as simple as possible so I don't think its worth splitting out for the puposes of this quick start.

mdpye replied:
I wanted to move it downwards as well, but I was cautioned against putting a `Drawable` property in the view type.

grzesiekko commented on app/Chatkit Quickstart/MessagesDataModel.swift:
```diff
@@ -59,22 +80,22 @@ class MessagesDataModel {
     }
     
-    public var delegate: MessagesDataModelDelegate?
+    var delegate: MessagesDataModelDelegate?
```
```suggestion
    weak var delegate: MessagesDataModelDelegate?
```

OliverPearmain-Pusher replied:
Addressed here (and in `ViewModel` also).  I have to make the `protocol`s `AnyObject` also.

grzesiekko commented on app/Chatkit Quickstart/OthersMessageTableViewCell.swift:
```diff
@@ -2,40 +2,52 @@ import UIKit
+        if let senderAvatarUrlString = senderAvatarUrl,
+            let senderAvatarUrl = URL(string: senderAvatarUrlString) {
+    
+            URLSession.shared.dataTask(with: senderAvatarUrl) { (data, _, _) in
```
Same as `SenderMessageTableViewCell.swift`.

grzesiekko commented on app/Chatkit Quickstart/MessagesViewModel.swift:
```diff
@@ -23,7 +28,7 @@ class MessagesViewModel: NSObject {
-    internal var items = [MessageView]()
+    private(set) var items = [MessageView]()
     
     var delegate: MessagesViewModelDelegate?
```
```suggestion
    weak var delegate: MessagesViewModelDelegate?
```

OliverPearmain-Pusher replied:
Sorted 👍

grzesiekko commented on app/Chatkit Quickstart/MessagesViewModel.swift:
```diff
@@ -8,7 +8,12 @@
 import Foundation
 
+protocol MessagesViewModelDelegate {
+    func didUpdate(model: [MessagesViewModel.MessageView], change: ChangeType)
```
An alternative approach that still uses `ChangeType`.

```suggestion
    func messagesViewModel(_ messagesViewModel: MessagesViewModel, didUpdate changeType: ChangeType)
```

OliverPearmain-Pusher replied:
Sorted.

OliverPearmain-Pusher committed:
Updates to delegate method signatures as suggested by Greg.  Also moved `extension MessagesViewModel: MessagesDataModelDelegate` into `MessagesViewModel.swift`.  Also removed some file header comments.

OliverPearmain-Pusher committed:
Fix compile issue from previous commit...
Updates to delegate method signatures as suggested by Greg.  Also moved `extension MessagesViewModel: MessagesDataModelDelegate` into `MessagesViewModel.swift`.  Also removed some file header comments.

OliverPearmain-Pusher committed:
Made `delegate` properties `weak` as per Greg's code review suggestion.

OliverPearmain-Pusher committed:
Removed now redundant `ChangeType`

OliverPearmain-Pusher committed:
Renamed a number of entities as below.  So now the `MessagesStore` vends a single `MessagesDataModel` which contains a list of `MessageDataItem`s.  And the `MessagesViewModel` vends a list of `MessageViewItem`s.

- MessagesDataModel -> MessagesStore
- MessagesDataModel.MessagesModel -> MessagesDataModel
- MessagesDataModel.MessageItem -> MessageDataItem
- MessagesViewModel.MessagesView -> MessagesViewItem

OliverPearmain-Pusher committed:
Added back in `DispatchQueue.main.async` because we cannot guarantee thread safety.

OliverPearmain-Pusher committed:
Merge `OKP/complete-revamp-history` into `OKP/complete-revamp-history-with-renames`

mdpye committed:
Merge pull request #9 from pusher/OKP/complete-revamp-history-with-renames

Renamed a number of entities.
